MongoDB-Liquibase does not pick up children ChangeLogs in native mode when using includeAll

### Describe the bug

When defining a changelog with a relative path linking to other changelogs, they are not found when launching a native application.

Given this hierarchy
```
.
└── resources/
    ├── db/
    │   ├── changeLogs/
    │   │   └── my-changeSet.yaml
    │   └── main-changeLog.yaml
    └── application.properties
```
where the `application.properties` defines `quarkus.liquibase-mongodb.change-log=db/main-changeLog.yaml` and where `main-changeLog.yaml` asks to retrieve changelogs from the `changeLogs/` directory
```
databaseChangeLog:
  - includeAll:
      relativeToChangelogFile: true
      path: changeLogs
      maxDepth: 1
```

`my-changeSet.yaml` will be correctly applied in JVM mode, while in native mode Liquibase will throw an error on startup
```
2025-03-25 19:05:58,595 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:121)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:77)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:48)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:137)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: java.lang.IllegalStateException: Error starting Liquibase
        at io.quarkus.liquibase.mongodb.runtime.LiquibaseMongodbRecorder.doStartActions(LiquibaseMongodbRecorder.java:73)
        at io.quarkus.runner.recorded.LiquibaseMongodbProcessor$startLiquibase1767457900.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.LiquibaseMongodbProcessor$startLiquibase1767457900.deploy(Unknown Source)
        ... 7 more
Caused by: liquibase.exception.CommandExecutionException: liquibase.exception.ChangeLogParseException: Error parsing db/main-changeLog.yaml : Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll 'changeLogs/'
        at liquibase.command.CommandScope.execute(CommandScope.java:258)
        at liquibase.Liquibase.lambda$validate$24(Liquibase.java:1247)
        at liquibase.Scope.lambda$child$0(Scope.java:191)
        at liquibase.Scope.child(Scope.java:200)
        at liquibase.Scope.child(Scope.java:190)
        at liquibase.Scope.child(Scope.java:169)
        at liquibase.Liquibase.runInScope(Liquibase.java:1290)
        at liquibase.Liquibase.validate(Liquibase.java:1242)
        at io.quarkus.liquibase.mongodb.runtime.LiquibaseMongodbRecorder.doStartActions(LiquibaseMongodbRecorder.java:63)
        ... 9 more
Caused by: liquibase.exception.ChangeLogParseException: Error parsing db/main-changeLog.yaml : Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll 'changeLogs/'
        at liquibase.parser.core.yaml.YamlChangeLogParser.parse(YamlChangeLogParser.java:98)
        at liquibase.command.core.helpers.DatabaseChangelogCommandStep.lambda$getDatabaseChangeLog$0(DatabaseChangelogCommandStep.java:129)
        at liquibase.Scope.child(Scope.java:200)
        at liquibase.Scope.child(Scope.java:176)
        at liquibase.command.core.helpers.DatabaseChangelogCommandStep.getDatabaseChangeLog(DatabaseChangelogCommandStep.java:128)
        at liquibase.command.core.helpers.DatabaseChangelogCommandStep.run(DatabaseChangelogCommandStep.java:87)
        at liquibase.command.CommandScope.execute(CommandScope.java:220)
        ... 17 more
Caused by: liquibase.exception.SetupException: Could not find directory, directory was empty, or no changelogs matched the provided search criteria for includeAll 'changeLogs/'
        at liquibase.changelog.DatabaseChangeLog.findResources(DatabaseChangeLog.java:879)
        at liquibase.changelog.DatabaseChangeLog.includeAll(DatabaseChangeLog.java:792)
        at liquibase.changelog.DatabaseChangeLog.handleChildNodeHelper(DatabaseChangeLog.java:540)
        at liquibase.changelog.DatabaseChangeLog.handleChildNode(DatabaseChangeLog.java:461)
        at liquibase.changelog.DatabaseChangeLog.load(DatabaseChangeLog.java:428)
        at liquibase.parser.core.yaml.YamlChangeLogParser.parse(YamlChangeLogParser.java:92)
        ... 23 more
```

I also added `quarkus.native.resources.includes = db/**`, but to no avail. As I understand it this is not important since the path resolution is done at build time.

Weirdly enough using `include` instead of `includeAll` to reference the file explicitly will work as expected in native mode:
```
databaseChangeLog:
  - include:
      file: changeLogs/changeSet.yaml
      relativeToChangelogFile: true
```

### How to Reproduce?

A reproducer can be found [here](https://github.com/celcius112/quarkus-mongo-liquibase) 

### Output of `uname -a` or `ver`

Darwin mac.lan 24.3.0 Darwin Kernel Version 24.3.0: Thu Jan  2 20:24:24 PST 2025; root:xnu-11215.81.4~3/RELEASE_ARM64_T6030 arm64

### Output of `java -version`

OpenJDK Runtime Environment GraalVM CE 21.0.2+13.1 (build 21.0.2+13-jvmci-23.1-b30)

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.19.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

gradle 3.12

--- a/independent-projects/bootstrap/runner/pom.xml
+++ b/independent-projects/bootstrap/runner/pom.xml
@@ -69,6 +69,11 @@
             <type>pom</type>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
 
         <dependency>
             <groupId>io.quarkus</groupId>

--- a/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java
+++ b/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java
@@ -93,18 +93,13 @@ public class JarResource implements ClassLoadingResource {
         private static final JarResourceURLProvider INSTANCE = new JarResourceURLProvider();
 
         @Override
-        public URL apply(JarFile jarFile, Path path, String res) {
-            JarEntry entry = jarFile.getJarEntry(res);
+        public URL apply(JarFile jarFile, Path path, String resource) {
+            JarEntry entry = jarFile.getJarEntry(resource);
             if (entry == null) {
                 return null;
             }
             try {
-                String realName = JarEntries.getRealName(entry);
-                // Avoid ending the URL with / to avoid breaking compatibility
-                if (realName.endsWith("/")) {
-                    realName = realName.substring(0, realName.length() - 1);
-                }
-                final URL resUrl = getUrl(path, realName);
+                final URL resUrl = getUrl(path, getRealName(entry, resource));
                 // wrap it up into a "jar" protocol URL
                 //horrible hack to deal with '?' characters in the URL
                 //seems to be the only way, the URI constructor just does not let you handle them in a sane way
@@ -122,6 +117,25 @@ public class JarResource implements ClassLoadingResource {
             } catch (MalformedURLException | URISyntaxException e) {
                 throw new RuntimeException(e);
             }
+        }
+
+        private static String getRealName(JarEntry entry, String resource) {
+            String realName = JarEntries.getRealName(entry);
+            // Make sure directories are returned with a / when the resource was requested with a /
+            if (resource.endsWith("/") && entry.isDirectory()) {
+                if (realName.endsWith("/")) {
+                    return realName;
+                } else {
+                    return realName + "/";
+                }
+            }
+
+            // this shouldn't be necessary but the previous implementation was doing it forcibly so keeping it for safety
+            if (realName.endsWith("/")) {
+                return realName.substring(0, realName.length() - 1);
+            }
+
+            return realName;
         }
 
         private static URL getUrl(Path jarPath, String realName) throws MalformedURLException, URISyntaxException {

--- a/independent-projects/bootstrap/runner/src/test/java/io/quarkus/bootstrap/runner/RunnerClassLoaderTest.java
+++ b/independent-projects/bootstrap/runner/src/test/java/io/quarkus/bootstrap/runner/RunnerClassLoaderTest.java
@@ -1,5 +1,6 @@
 package io.quarkus.bootstrap.runner;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.nio.file.Path;
@@ -96,6 +97,27 @@ public class RunnerClassLoaderTest {
         }
 
         assertTrue(exceptionsInThreads.isEmpty(), "Exceptions in threads: " + exceptionsInThreads);
+    }
+
+    @Test
+    public void testUrlWithTrailingSlash() {
+        ClassLoadingResource[] classLoadingResources = new ClassLoadingResource[] {
+                createProjectJarResource("simple-project-1.0.jar") };
+
+        Map<String, ClassLoadingResource[]> resourceDirectoryMap = Map.of(
+                "org", classLoadingResources,
+                "org/simple", classLoadingResources);
+
+        RunnerClassLoader runnerClassLoader = new RunnerClassLoader(ClassLoader.getSystemClassLoader(), resourceDirectoryMap,
+                Collections.emptySet(), Collections.emptySet(),
+                Collections.emptyList(), Collections.emptyMap());
+
+        assertThat(runnerClassLoader.findResource("org").toString()).endsWith("/org");
+        assertThat(runnerClassLoader.findResource("org/").toString()).endsWith("/org/");
+        assertThat(runnerClassLoader.findResource("org/simple").toString()).endsWith("/org/simple");
+        assertThat(runnerClassLoader.findResource("org/simple/").toString()).endsWith("/org/simple/");
+        assertThat(runnerClassLoader.findResource("org/simple/SimplePojo1.class").toString())
+                .endsWith("/org/simple/SimplePojo1.class");
     }
 
     private static JarResource createProjectJarResource(String jarName) {

